homework-assignments-1/task4.ts

```diff
@@ -0,0 +1,179 @@
+class Product {
+  constructor(
+    public name: string,
+    public price: number,
+    public stock: number
+  ) {}
+
+  addStock(quantity: number): void {
+    this.stock += quantity;
+  }
+
+  reduceStock(quantity: number): void {
+    if (this.stock >= quantity) {
+      this.stock -= quantity;
+    } else {
+      console.log(`Not enough stock for ${this.name}`);
+    }
+  }
+}
+
+class Customer {
+  public cart: Product[] = [];
+
+  constructor(
+    public name: string,
+    public email: string,
+    public address: string
+  ) {}
+
+  addToCart(product: Product): void {
+    this.cart.push(product);
+    console.log(`${product.name} added to cart`);
+  }
+
+  removeFromCart(product: Product): void {
+    this.cart = this.cart.filter((p) => p !== product);
+    console.log(`${product.name} removed from cart`);
+  }
+}
+
+abstract class Order {
+  public totalPrice: number = 0;
+
+  constructor(public customer: Customer, public products: Product[]) {}
+
+  abstract calculateTotalPrice(): void;
+}
+class PhysicalOrder extends Order {
+  public shippingAddress: string;
+  public deliveryDate: Date;
+
+  constructor(
+    customer: Customer,
+    products: Product[],
+    shippingAddress: string
+  ) {
+    super(customer, products);
+    this.shippingAddress = shippingAddress;
+    this.deliveryDate = new Date();
+    this.deliveryDate.setDate(this.deliveryDate.getDate() + 7); // Estimate delivery in 7 days
+  }
+
+  calculateTotalPrice(): void {
+    const productTotal = this.products.reduce(
+      (sum, product) => sum + product.price,
```

Comment:
it is not good to access object member properties directly. only setters and getters should interact with them